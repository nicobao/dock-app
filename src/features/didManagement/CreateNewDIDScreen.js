import React, {useMemo, useRef, useState} from 'react';
import {
  BackButton,
  Box,
  Header,
  IconButton,
  LoadingButton,
  NBox,
  Input,
  ScreenContainer,
  Theme,
  Typography,
} from '../../design-system';
import {navigateBack} from '../../core/navigation';
import {
  FormControl,
  HStack,
  Icon,
  ScrollView,
  Stack,
  VStack,
} from 'native-base';
import {translate} from '../../locales';
import {Ionicons} from '@native-base/icons';
import {addTestId} from '../../core/automation-utils';
import {useDIDManagementHandlers} from './didHooks';
import QuickInfoIcon from '../../assets/icons/quick-info.svg';
import {CustomSelectInput} from '../../components/CustomSelectInput';
import {useAccountsList} from '../accounts/accountsHooks';
import {CreateDIDDockConfirmationModal} from './components/CreateDIDDockConfirmationModal';

export function CreateNewDIDScreen({
  form,
  handleChange,
  handleSubmit,
  accounts,
  isFormValid,
}) {
  const [isConfirmDIDDockVisible, setIsConfirmDIDDockVisible] = useState(false);
  const scrollViewRef = useRef();

  return (
    <ScreenContainer {...addTestId('CreateNewDIDScreen')}>
      <Header>
        <Box
          marginLeft={1}
          marginRight={22}
          flexDirection="row"
          alignItems="center">
          <NBox width={'80px'}>
            <BackButton
              {...addTestId('CreateNewDIDScreenGoBack')}
              onPress={navigateBack}
            />
          </NBox>
          <NBox
            flex={1}
            width="100%"
            alignContent="center"
            alignItems="center"
            pl={15}>
            <Typography variant="h3">
              {translate('didManagement.create_new_did')}
            </Typography>
          </NBox>
          <NBox width="80px" alignItems="flex-end" />
        </Box>
      </Header>
      <ScrollView
        ref={view => {
          scrollViewRef.current = view;
        }}
        marginLeft={3}
        marginRight={3}>
        <FormControl>
          <Stack mt={7}>
            <FormControl.Label>
              {translate('didManagement.did_name')}
            </FormControl.Label>
            <Input
              {...addTestId('DIDName')}
              value={form.didName}
              onChangeText={handleChange('didName')}
              autoCapitalize="none"
            />
            <FormControl.HelperText>
              {translate('didManagement.did_name_input_hint')}
            </FormControl.HelperText>
          </Stack>
        </FormControl>

        <FormControl isInvalid={form._errors.didType}>
          <Stack mt={7}>
            <FormControl.Label>
              {translate('didManagement.did_type')}
            </FormControl.Label>
            <CustomSelectInput
              onPressItem={item => {
                handleChange('didType')(item.value);
                if (item.value === 'didkey') {
                  handleChange('didPaymentAddress')('');
                }
              }}
              renderItem={item => {
                return (
                  <>
                    <Typography textAlign="left" variant="description">
                      {item.label}
                    </Typography>
                    <Typography textAlign="left" variant="screen-description">
                      {item.description}
                    </Typography>
                  </>
                );
              }}
              items={[
                {
                  value: 'didkey',
                  label: translate('didManagement.did_key'),
                  description: translate('didManagement.did_key_description'),
                },
                {
                  value: 'diddock',
                  label: translate('didManagement.did_dock'),
                  description: translate('didManagement.did_dock_description'),
                },
              ]}
            />
          </Stack>
          <FormControl.ErrorMessage>
            {form._errors.didType}
          </FormControl.ErrorMessage>
        </FormControl>

        {form.didType === 'diddock' ? (
          <FormControl isInvalid={form._errors.didPaymentAddress}>
            <Stack mt={7}>
              <FormControl.Label>
                {translate('didManagement.did_payment_address')}
              </FormControl.Label>
              <CustomSelectInput
                emptyItemMessage={translate('account_details.no_account_exist')}
                onPressItem={item => {
                  handleChange('didPaymentAddress')(item.value);
                }}
                renderItem={item => {
                  return (
                    <>
                      <Typography
                        numberOfLines={1}
                        textAlign="left"
                        variant="description">
                        {item.label}
                      </Typography>
                      <Typography
                        numberOfLines={1}
                        textAlign="left"
                        variant="screen-description">
                        {item.description}
                      </Typography>
                    </>
                  );
                }}
                items={accounts}
              />
            </Stack>
            <FormControl.ErrorMessage>
              {form._errors.didPaymentAddress}
            </FormControl.ErrorMessage>
          </FormControl>
        ) : null}

        {form.didType === 'diddock' && form.showDIDDockQuickInfo ? (
          <VStack
            mt={7}
            px={5}
            pt={3}
            pb={5}
            style={{
              backgroundColor: Theme.colors.bgBlue,
              borderRadius: Theme.borderRadius,
            }}>
            <HStack>
              <HStack
                style={{
                  flexGrow: 1,
                }}>
                <NBox mt={4}>
                  <QuickInfoIcon />
                </NBox>
                <Typography mt={2} ml={1} variant="h3">
                  {' ' + translate('didManagement.quick_info')}
                </Typography>
              </HStack>
              <IconButton
                {...addTestId('CreateNewDIDScreenCloseQuickInfo')}
                onPress={() => {
                  handleChange('showDIDDockQuickInfo')(false);
                }}>
                <Icon as={Ionicons} name="close-outline" />
              </IconButton>
            </HStack>
            <Typography textAlign="left" variant="didDescription">
              {translate('didManagement.did_dock_info')}
            </Typography>
          </VStack>
        ) : null}
      </ScrollView>
      <LoadingButton
        full
        {...addTestId('CreateNewDIDScreenDIDCreate')}
        mb={30}
        ml={3}
        mr={3}
        onPress={async () => {
          if (form.didType === 'didkey') {
            await handleSubmit();
          } else if (form.didType === 'diddock') {
            setIsConfirmDIDDockVisible(true);
          }
        }}
        isDisabled={!isFormValid}>
        {translate('didManagement.create')}
      </LoadingButton>
      <CreateDIDDockConfirmationModal
        onClose={() => {
          setIsConfirmDIDDockVisible(false);
        }}
        visible={isConfirmDIDDockVisible}
        didName={form.didName}
        didType={form.didType}
        onCreateDID={handleSubmit}
      />
    </ScreenContainer>
  );
}
export function CreateNewDIDScreenContainer() {
  const {form, onCreateDID, handleChange, isFormValid} =
    useDIDManagementHandlers();
  const {accounts} = useAccountsList();

  const parseAccounts = useMemo(() => {
    return accounts.map(account => {
      return {
        value: account.address,
        label: account.name,
        description: account.address,
      };
    });
  }, [accounts]);

  return (
    <CreateNewDIDScreen
      isFormValid={isFormValid}
      handleChange={handleChange}
      form={form}
      handleSubmit={onCreateDID}
      accounts={parseAccounts}
    />
  );
}
