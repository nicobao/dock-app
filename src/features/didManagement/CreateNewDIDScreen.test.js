import React from 'react';
import {shallow} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import {CreateNewDIDScreen} from './CreateNewDIDScreen';

const mockStore = configureMockStore();

const accounts = [
  {
    value: '6GwnHZARcEkJio9dxPYy6SC5sAL6PxpZAB6VYwoFjGMU',
    label: 'Test Account',
    description: '6GwnHZARcEkJio9dxPYy6SC5sAL6PxpZAB6VYwoFjGMU',
  },
];
describe('CreateNewDIDScreen', () => {
  it('should render correctly when form is invalid', () => {
    const initialState = {
      // placeholder for redux store
    };

    const form = {
      didName: '',
      didType: 'didkey',
      showDIDDockQuickInfo: false,
      keypairType: 'ed25519',
      derivationPath: '',
      _errors: {
        didType: '',
      },
      _hasError: false,
    };
    const wrapper = shallow(
      <CreateNewDIDScreen
        isFormValid={false}
        accounts={[]}
        handleChange={jest.fn()}
        form={form}
        handleSubmit={jest.fn()}
      />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('should render correctly when form is valid', () => {
    const initialState = {
      // placeholder for redux store
    };

    const form = {
      didName: 'Default',
      didType: 'didkey',
      showDIDDockQuickInfo: false,
      keypairType: 'ed25519',
      derivationPath: '',
      _errors: {
        didType: '',
      },
      _hasError: false,
    };
    const wrapper = shallow(
      <CreateNewDIDScreen
        isFormValid={true}
        accounts={[]}
        handleChange={jest.fn()}
        form={form}
        handleSubmit={jest.fn()}
      />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('should render correctly with modal', () => {
    const initialState = {
      // placeholder for redux store
    };

    const form = {
      didName: '',
      didType: 'diddock',
      showDIDDockQuickInfo: true,
      keypairType: 'ed25519',
      derivationPath: '',
      _errors: {
        didType: '',
      },
      _hasError: false,
    };
    const wrapper = shallow(
      <CreateNewDIDScreen
        isFormValid={true}
        accounts={accounts}
        handleChange={jest.fn()}
        form={form}
        handleSubmit={jest.fn()}
      />,
      {
        context: {store: mockStore(initialState)},
      },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
