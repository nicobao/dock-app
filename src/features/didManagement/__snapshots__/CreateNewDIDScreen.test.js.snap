// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewDIDScreen should render correctly when form is invalid 1`] = `
<RNCSafeAreaView
  accessibilityLabel="CreateNewDIDScreen"
  style={
    Object {
      "backgroundColor": "#141417",
      "flex": 1,
    }
  }
  testID="CreateNewDIDScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <Styled(View)>
        <Box
          alignItems="center"
          flexDirection="row"
          marginLeft={1}
          marginRight={22}
        >
          <Memo(ForwardRef(Box))
            width="80px"
          >
            <BackButton
              accessibilityLabel="CreateNewDIDScreenGoBack"
              onPress={[MockFunction]}
              testID="CreateNewDIDScreenGoBack"
            />
          </Memo(ForwardRef(Box))>
          <Memo(ForwardRef(Box))
            alignContent="center"
            alignItems="center"
            flex={1}
            pl={15}
            width="100%"
          >
            <Styled(Component)
              variant="h3"
            >
              Create New DID
            </Styled(Component)>
          </Memo(ForwardRef(Box))>
          <Memo(ForwardRef(Box))
            alignItems="flex-end"
            width="80px"
          />
        </Box>
      </Styled(View)>
      <ForwardRef
        marginLeft={3}
        marginRight={3}
      >
        <Memo(ForwardRef(FormControl))>
          <Memo(ForwardRef(Stack))
            mt={7}
          >
            <Memo(ForwardRef(FormControlLabel))>
              DID Name
            </Memo(ForwardRef(FormControlLabel))>
            <Input
              accessibilityLabel="DIDName"
              autoCapitalize="none"
              testID="DIDName"
              value=""
            />
            <Memo(ForwardRef(FormControlHelperText))>
              This will just be used for easy reference
            </Memo(ForwardRef(FormControlHelperText))>
          </Memo(ForwardRef(Stack))>
        </Memo(ForwardRef(FormControl))>
        <Memo(ForwardRef(FormControl))
          isInvalid=""
        >
          <Memo(ForwardRef(Stack))
            mt={7}
          >
            <Memo(ForwardRef(FormControlLabel))>
              DID Type
            </Memo(ForwardRef(FormControlLabel))>
            <CustomSelectInput
              items={
                Array [
                  Object {
                    "description": "W3C compliant Decentralized identifier",
                    "label": "did:key",
                    "value": "didkey",
                  },
                  Object {
                    "description": "W3C compliant Decentralized Identifier created by Dock. Uses blockchain to secure verifiable credentials.",
                    "label": "did:dock",
                    "value": "diddock",
                  },
                ]
              }
              onPressItem={[Function]}
              renderItem={[Function]}
            />
          </Memo(ForwardRef(Stack))>
          <Memo(ForwardRef(FormControlErrorMessage)) />
        </Memo(ForwardRef(FormControl))>
        <DIDAdvancedOptions
          accessibilityLabel="DIDAdvancedOptions"
          form={
            Object {
              "_errors": Object {
                "didType": "",
              },
              "_hasError": false,
              "derivationPath": "",
              "didName": "",
              "didType": "didkey",
              "keypairType": "ed25519",
              "showDIDDockQuickInfo": false,
            }
          }
          onChange={
            [MockFunction] {
              "calls": Array [
                Array [
                  "didName",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          scrollToBottom={[Function]}
          testID="DIDAdvancedOptions"
        />
      </ForwardRef>
      <LoadingButton
        accessibilityLabel="CreateNewDIDScreenDIDCreate"
        full={true}
        isDisabled={true}
        mb={30}
        ml={3}
        mr={3}
        onPress={[Function]}
        testID="CreateNewDIDScreenDIDCreate"
      >
        Create
      </LoadingButton>
      <CreateDIDDockConfirmationModal
        didName=""
        didType="didkey"
        onClose={[Function]}
        onCreateDID={[MockFunction]}
        visible={false}
      />
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;

exports[`CreateNewDIDScreen should render correctly when form is valid 1`] = `
<RNCSafeAreaView
  accessibilityLabel="CreateNewDIDScreen"
  style={
    Object {
      "backgroundColor": "#141417",
      "flex": 1,
    }
  }
  testID="CreateNewDIDScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <Styled(View)>
        <Box
          alignItems="center"
          flexDirection="row"
          marginLeft={1}
          marginRight={22}
        >
          <Memo(ForwardRef(Box))
            width="80px"
          >
            <BackButton
              accessibilityLabel="CreateNewDIDScreenGoBack"
              onPress={[MockFunction]}
              testID="CreateNewDIDScreenGoBack"
            />
          </Memo(ForwardRef(Box))>
          <Memo(ForwardRef(Box))
            alignContent="center"
            alignItems="center"
            flex={1}
            pl={15}
            width="100%"
          >
            <Styled(Component)
              variant="h3"
            >
              Create New DID
            </Styled(Component)>
          </Memo(ForwardRef(Box))>
          <Memo(ForwardRef(Box))
            alignItems="flex-end"
            width="80px"
          />
        </Box>
      </Styled(View)>
      <ForwardRef
        marginLeft={3}
        marginRight={3}
      >
        <Memo(ForwardRef(FormControl))>
          <Memo(ForwardRef(Stack))
            mt={7}
          >
            <Memo(ForwardRef(FormControlLabel))>
              DID Name
            </Memo(ForwardRef(FormControlLabel))>
            <Input
              accessibilityLabel="DIDName"
              autoCapitalize="none"
              testID="DIDName"
              value="Default"
            />
            <Memo(ForwardRef(FormControlHelperText))>
              This will just be used for easy reference
            </Memo(ForwardRef(FormControlHelperText))>
          </Memo(ForwardRef(Stack))>
        </Memo(ForwardRef(FormControl))>
        <Memo(ForwardRef(FormControl))
          isInvalid=""
        >
          <Memo(ForwardRef(Stack))
            mt={7}
          >
            <Memo(ForwardRef(FormControlLabel))>
              DID Type
            </Memo(ForwardRef(FormControlLabel))>
            <CustomSelectInput
              items={
                Array [
                  Object {
                    "description": "W3C compliant Decentralized identifier",
                    "label": "did:key",
                    "value": "didkey",
                  },
                  Object {
                    "description": "W3C compliant Decentralized Identifier created by Dock. Uses blockchain to secure verifiable credentials.",
                    "label": "did:dock",
                    "value": "diddock",
                  },
                ]
              }
              onPressItem={[Function]}
              renderItem={[Function]}
            />
          </Memo(ForwardRef(Stack))>
          <Memo(ForwardRef(FormControlErrorMessage)) />
        </Memo(ForwardRef(FormControl))>
        <DIDAdvancedOptions
          accessibilityLabel="DIDAdvancedOptions"
          form={
            Object {
              "_errors": Object {
                "didType": "",
              },
              "_hasError": false,
              "derivationPath": "",
              "didName": "Default",
              "didType": "didkey",
              "keypairType": "ed25519",
              "showDIDDockQuickInfo": false,
            }
          }
          onChange={
            [MockFunction] {
              "calls": Array [
                Array [
                  "didName",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          scrollToBottom={[Function]}
          testID="DIDAdvancedOptions"
        />
      </ForwardRef>
      <LoadingButton
        accessibilityLabel="CreateNewDIDScreenDIDCreate"
        full={true}
        isDisabled={false}
        mb={30}
        ml={3}
        mr={3}
        onPress={[Function]}
        testID="CreateNewDIDScreenDIDCreate"
      >
        Create
      </LoadingButton>
      <CreateDIDDockConfirmationModal
        didName="Default"
        didType="didkey"
        onClose={[Function]}
        onCreateDID={[MockFunction]}
        visible={false}
      />
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;

exports[`CreateNewDIDScreen should render correctly with modal 1`] = `
<RNCSafeAreaView
  accessibilityLabel="CreateNewDIDScreen"
  style={
    Object {
      "backgroundColor": "#141417",
      "flex": 1,
    }
  }
  testID="CreateNewDIDScreen"
>
  <AppGlobalHeader />
  <Memo(ForwardRef(Box))
    flex={1}
  >
    <Memo(ForwardRef(Box))
      flex={1}
    >
      <Styled(View)>
        <Box
          alignItems="center"
          flexDirection="row"
          marginLeft={1}
          marginRight={22}
        >
          <Memo(ForwardRef(Box))
            width="80px"
          >
            <BackButton
              accessibilityLabel="CreateNewDIDScreenGoBack"
              onPress={[MockFunction]}
              testID="CreateNewDIDScreenGoBack"
            />
          </Memo(ForwardRef(Box))>
          <Memo(ForwardRef(Box))
            alignContent="center"
            alignItems="center"
            flex={1}
            pl={15}
            width="100%"
          >
            <Styled(Component)
              variant="h3"
            >
              Create New DID
            </Styled(Component)>
          </Memo(ForwardRef(Box))>
          <Memo(ForwardRef(Box))
            alignItems="flex-end"
            width="80px"
          />
        </Box>
      </Styled(View)>
      <ForwardRef
        marginLeft={3}
        marginRight={3}
      >
        <Memo(ForwardRef(FormControl))>
          <Memo(ForwardRef(Stack))
            mt={7}
          >
            <Memo(ForwardRef(FormControlLabel))>
              DID Name
            </Memo(ForwardRef(FormControlLabel))>
            <Input
              accessibilityLabel="DIDName"
              autoCapitalize="none"
              testID="DIDName"
              value=""
            />
            <Memo(ForwardRef(FormControlHelperText))>
              This will just be used for easy reference
            </Memo(ForwardRef(FormControlHelperText))>
          </Memo(ForwardRef(Stack))>
        </Memo(ForwardRef(FormControl))>
        <Memo(ForwardRef(FormControl))
          isInvalid=""
        >
          <Memo(ForwardRef(Stack))
            mt={7}
          >
            <Memo(ForwardRef(FormControlLabel))>
              DID Type
            </Memo(ForwardRef(FormControlLabel))>
            <CustomSelectInput
              items={
                Array [
                  Object {
                    "description": "W3C compliant Decentralized identifier",
                    "label": "did:key",
                    "value": "didkey",
                  },
                  Object {
                    "description": "W3C compliant Decentralized Identifier created by Dock. Uses blockchain to secure verifiable credentials.",
                    "label": "did:dock",
                    "value": "diddock",
                  },
                ]
              }
              onPressItem={[Function]}
              renderItem={[Function]}
            />
          </Memo(ForwardRef(Stack))>
          <Memo(ForwardRef(FormControlErrorMessage)) />
        </Memo(ForwardRef(FormControl))>
        <Memo(ForwardRef(FormControl))>
          <Memo(ForwardRef(Stack))
            mt={7}
          >
            <Memo(ForwardRef(FormControlLabel))>
              DID Payment Account
            </Memo(ForwardRef(FormControlLabel))>
            <CustomSelectInput
              emptyItemMessage="No account exists"
              items={
                Array [
                  Object {
                    "description": "6GwnHZARcEkJio9dxPYy6SC5sAL6PxpZAB6VYwoFjGMU",
                    "label": "Test Account",
                    "value": "6GwnHZARcEkJio9dxPYy6SC5sAL6PxpZAB6VYwoFjGMU",
                  },
                ]
              }
              onPressItem={[Function]}
              renderItem={[Function]}
            />
          </Memo(ForwardRef(Stack))>
          <Memo(ForwardRef(FormControlErrorMessage)) />
        </Memo(ForwardRef(FormControl))>
        <Memo(ForwardRef(VStack))
          mt={7}
          pb={5}
          pt={3}
          px={5}
          style={
            Object {
              "backgroundColor": "#00174D",
              "borderRadius": 6,
            }
          }
        >
          <Memo(ForwardRef(HStack))>
            <Memo(ForwardRef(HStack))
              style={
                Object {
                  "flexGrow": 1,
                }
              }
            >
              <Memo(ForwardRef(Box))
                mt={4}
              >
                <SvgFile />
              </Memo(ForwardRef(Box))>
              <Styled(Component)
                ml={1}
                mt={2}
                variant="h3"
              >
                 Quick Info
              </Styled(Component)>
            </Memo(ForwardRef(HStack))>
            <IconButton
              accessibilityLabel="CreateNewDIDScreenCloseQuickInfo"
              onPress={[Function]}
              testID="CreateNewDIDScreenCloseQuickInfo"
            >
              <Memo(ForwardRef(Icon))
                as={[Function]}
                name="close-outline"
              />
            </IconButton>
          </Memo(ForwardRef(HStack))>
          <Styled(Component)
            textAlign="left"
            variant="didDescription"
          >
            did:dock requires DOCK token to create. This is used for the transaction fee required to create the DID and have a record on the blockchain.
          </Styled(Component)>
        </Memo(ForwardRef(VStack))>
        <DIDAdvancedOptions
          accessibilityLabel="DIDAdvancedOptions"
          form={
            Object {
              "_errors": Object {
                "didType": "",
              },
              "_hasError": false,
              "derivationPath": "",
              "didName": "",
              "didType": "diddock",
              "keypairType": "ed25519",
              "showDIDDockQuickInfo": true,
            }
          }
          onChange={
            [MockFunction] {
              "calls": Array [
                Array [
                  "didName",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          scrollToBottom={[Function]}
          testID="DIDAdvancedOptions"
        />
      </ForwardRef>
      <LoadingButton
        accessibilityLabel="CreateNewDIDScreenDIDCreate"
        full={true}
        isDisabled={false}
        mb={30}
        ml={3}
        mr={3}
        onPress={[Function]}
        testID="CreateNewDIDScreenDIDCreate"
      >
        Create
      </LoadingButton>
      <CreateDIDDockConfirmationModal
        didName=""
        didType="diddock"
        onClose={[Function]}
        onCreateDID={[MockFunction]}
        visible={false}
      />
    </Memo(ForwardRef(Box))>
  </Memo(ForwardRef(Box))>
</RNCSafeAreaView>
`;
